Added the ability to multiply a SparseTensor with a Dense Matrix on the left side

joglekara opened the pull request:
At the moment, functionality for `sp_a x dense_b` is implemented. 

It is easy to show using some linear algebra identities that `dense_a x sp_b` can be acquired by multiplication of the adjoints of the two constituent matrices, and transposing the result.

We implement this so that future users do not have to prove to themselves that the implementations of the above operations in TensorFlow are adequate, and that the linear algebra techniques may apply.

This enables those users who wish to compute `dense_a x sp_b` to quickly do so.

Commit:
Added dense_sparse_matmul()

This commit adds tf.SparseTensor.dense_sparse_matmul().

It uses sparse_dense_matmul() and some matrix manipulation to get to the right answer.

It has been tested against numpy's calculations of the same operations

Commit:
Merge remote-tracking branch 'upstream/master' into dense_sparse_matmul

This was performed in order to pull the upstream changes (e.g. v2.0) into the branch before creating a pull request.

Commit:
Updating to tf2.0 specs and added a test

extremely minor changes include
1 - renaming the function calls

also added a test
manual construction of a sparse and dense tensor and checking that the result of their multiplication is the same as that given by `tf.matmul`

alextp commented on tensorflow/python/ops/sparse_ops.py:
```diff
@@ -2404,6 +2404,34 @@ def sparse_tensor_dense_matmul(sp_a,
+           v1=["sparse.dense_sparse_matmul",
+               "dense_sparse_tensor_matmul"])
+@deprecation.deprecated_endpoints("dense_sparse_tensor_matmul")
+def dense_sparse_tensor_matmul(a,
```
Isn't it better to extend the existing sparse_dense_matmul to support the other order of arguments?

joglekara replied:
hmm, that's a great suggestion. 

That will actually extend the functionality even further. I will update accordingly!

joglekara replied:
do you have a recommendation for how to implement a check that determines whether `a` is sparse or `b`?

alextp commented:
isinstance(a, SparseTensor)?

On Mon, Dec 2, 2019 at 2:08 PM Archis Joglekar <notifications@github.com>
wrote:

> *@joglekara* commented on this pull request.
> ------------------------------
>
> In tensorflow/python/ops/sparse_ops.py
> <https://github.com/tensorflow/tensorflow/pull/34624#discussion_r352885008>
> :
>
> > @@ -2404,6 +2404,34 @@ def sparse_tensor_dense_matmul(sp_a,
>          adjoint_a=adjoint_a,
>          adjoint_b=adjoint_b)
>
> +tf_export("sparse.dense_sparse_matmul",
> +           v1=["sparse.dense_sparse_matmul",
> +               "dense_sparse_tensor_matmul"])
> +@deprecation.deprecated_endpoints("dense_sparse_tensor_matmul")
> +def dense_sparse_tensor_matmul(a,
>
> do you have a recommendation for how to implement a check that determines
> whether a is sparse or b?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/34624?email_source=notifications&email_token=AAABHRP6YSZ5X25FI2DSRGDQWWBOPA5CNFSM4JR5UYS2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCNVPH6I#discussion_r352885008>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRNU6SEZ66Y3AGFY7KTQWWBOPANCNFSM4JR5UYSQ>
> .
>


-- 
 - Alex

Commit:
Extend the existing sparse_dense_matmul to support the other order of arguments

Implement type checking to see which matrix is sparse and which isn't.
Implement checking for `adjoint_a/b` flags because that affects the implementation of the reverse-order multiplication.
This adds functionality to the previous commit by allowing any permutation of sparse, dense, adjoint_a, adjoint_b

joglekara commented:
Thanks, have merged the two functions as recommended, and that generalization has led to added  functionality.

Commit:
Updating the tests

This tests fine locally with `bazel run //tensorflow/python:sparse_ops_test`

joglekara commented:
Hi @alextp, I had to update the tests. This commit passes `bazel run //tensorflow/python:sparse_ops_test` locally.

joglekara commented:
hey @alextp , I had to update a few of the tests. Nothing's changed in terms of the actual implementation. Please let me know if you have any further comments or thoughts.

Commit:
Include check for SparseTensorValue

Previous iteration was only checking for SparseTensor previously.

Was failing a test when a SparseTensorValue was fed in.

`bazel run //tensorflow/python/kernel_tests:sparse_tensor_dense_matmul_op_test` passes now.

Commit:
Merge remote-tracking branch 'upstream/master' into dense_sparse_matmul
Doing this in hopes that it fixes the error related to the API compatibility test

gbaned commented:
@joglekara Can you please address Ubuntu Sanity errors? Thanks!

joglekara commented:
@gbaned It's an `api_compatability_test` error. I'm working on reproducing the error locally. If you have any thoughts on what might help, please let me know.

joglekara commented:
Update: I've found the instructions to run the Sanity test, and it is running now.

Commit:
lines-too-long

reformatted to address the 80 character limit on a line.

joglekara commented:
hi @alextp, again, no code changes. The unit tests are good to go as well, it is just style this time. This commit addressed some `line-too-long` warnings from `pylint`. 

Hope this passes Ubuntu Sanity. If not, I am running the Sanity test locally in the meantime and should be able to get this through soon.

joglekara commented:
Thanks for pushing that through. Looks like Sanity is fine now, but `Ubuntu CPU` is still having issues with the `API_compatibility_test`. 
It would be useful if it could show where the incompatibility was in addition to how many. Anyway, I'm on it!

Also, if anyone could tell me what went wrong with `import/copybara`, I can try to address that as well

Commit:
Updated golden API according to test

For reference:
API has been modified due to change in name of variables.
 The new variable names reflect new capability where either input can be sparse or dense. The pre-existing variable names became misleading with this added functionality.

joglekara commented:
I got to the bottom of the `api_compatibility_test` problems and have updated the golden labels accordingly. 

Please review the new changes and let me know.

For reference:
The API has been modified due to a change in the name of variables. The new variable names reflect the fact that either input can be sparse or dense. The pre-existing variable names became misleading with this added functionality.

alextp reviewed (changes requested):
The API changes you propose are not backwards compatible. If someone was calling this function passing keyword arguments their code would break. Please either revert the name change (strongly preferred) and document the new meaning or add new arguments with the correct name at the end, make the old names optional, and mark them as deprecated arguments (but, really, the other one)

joglekara commented:
Thanks for the feedback, makes sense. I will choose the former.

Can you tell me exactly where to document the new meaning? At least the `docstring`, but is there anywhere else this should go?

alextp commented:
Just the docstring.

On Thu, Jan 2, 2020 at 8:58 AM Archis Joglekar <notifications@github.com>
wrote:

> Thanks for the feedback, makes sense. I will choose the former.
>
> Can you tell me exactly where to document the new meaning? At least the
> docstring, but is there anywhere else this should go?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/34624?email_source=notifications&email_token=AAABHRNRAMAKFLH2IOX73PLQ3YMMNA5CNFSM4JR5UYS2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH6Z6TQ#issuecomment-570269518>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABHRMYW4UTEY2DSKETKY3Q3YMMNANCNFSM4JR5UYSQ>
> .
>


-- 
 - Alex

Commit:
Reset API and corresponding golden and added documentation to docstring.

joglekara commented:
Thanks. Just pushed the requested changes.

joglekara commented:
@rthadur, any idea on what the `copybara` error might be?

mihaimaruseac commented:
Can you please rebase on master?

Commit:
Merge remote-tracking branch 'upstream/master' into dense_sparse_matmul

joglekara commented:
Done

joglekara commented:
@mihaimaruseac @rthadur please let me know if there's something else you need me to do on my side. this is quite mysterious to me!

mihaimaruseac commented:
Windows failures are unrelated, so let's wait for internal review and import + additional testing after that

joglekara commented:
do you mean @alextp or is someone else internal taking a look at it?

mihaimaruseac commented:
All code from GitHub gets imported internally at Google, integrated with existing code there and reviewed again. There are more tests being run there.

joglekara commented:
hey @mihaimaruseac, let me know if there is any feedback for me that I can use to improve my submission. Thanks!

mihaimaruseac commented:
I think this is good to go, but internal testing hasn't been run yet due to some tooling issue. Sorry about that.

joglekara commented:
Ah alright. Do you happen to have an idea of when that might be resolved?

mihaimaruseac commented:
The tooling issues have just been resolved. So we're going through the pipeline of blocked PRs and merging them

joglekara commented:
Great. Thanks for the update!

joglekara commented:
Looks like something's getting stuck, let me know if there's something I can do to help!

mihaimaruseac commented on tensorflow/python/ops/sparse_ops.py:
```diff
@@ -2392,18 +2394,39 @@ def sparse_tensor_dense_matmul(sp_a,
+  if isinstance(b, sparse_tensor.SparseTensor) \
+          or isinstance(b, sparse_tensor.SparseTensorValue):
+
+    if adjoint_a == True and adjoint_b == False:
```
`Comparison to False should be 'not expr' or 'expr is False' [singleton-comparison]`

mihaimaruseac replied:
`Comparison to True should be just 'expr' or 'expr is True'`

mihaimaruseac replied:
`Use implicit True/False evaluation`

mihaimaruseac reviewed (approved):
So the change couldn't be approved due to style issues. Will fix it internally

mihaimaruseac commented on tensorflow/python/ops/sparse_ops.py:
```diff
@@ -2392,18 +2394,39 @@ def sparse_tensor_dense_matmul(sp_a,
+    elif adjoint_a == True and adjoint_b == True:
+      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
+                                                            adjoint_a=False,
+                                                            adjoint_b=False))
```
Actually, we need you to fix this. Is it intentional that the branches are changed here? (testing for both `andjoint_* == False` but setting them to `True` when calling the function)

If they have to be the same, why not just use

```
return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a, adjoin_a, adjoin_b)
```

mihaimaruseac replied:
Don't forget about this change please. Is this intentional or a typo?

joglekara replied:
Yes, we have to reverse the `adjoint_*` flags when you reverse the order of the multiply. 

To explain:

We have functionality for CD where C is sparse. To get the product AB where B is sparse, we calculate:

`AB = (B.T A.T).T `

This above format puts its back in the form that we have, namely, CD where C is sparse. Here, C now represents B.T, and D represents A.T.

So calculating AB where B is sparse requires the highest level call to have `adjoint_* = False`, while the sub-call, with the order swapped, requires them both to be true. That subcall also has to wrapped with the `transpose` op as you noted. 

All of this is to say that I think this is the most succinct way I can write the above logic, and it passes the tests to boot.

Does this make sense? Am I missing something obvious?

joglekara replied:
Okay to resolve this segment?

mihaimaruseac replied:
Oh, I just saw you're flipping the arguments. Now it makes more sense.

How about rewriting the entire code from line 2401 to 2417 like:

```python
    # We can do C * D where C is sparse but if we want to do A * B when
    # B is sparse we have to transpose. But AB = (B'A')' so we have to transpose.
    if adjoint_a != adjoint_b:
￼      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a, adjoint_a, adjoint_b))
￼    else:
￼      return array_ops.transpose(sparse_tensor_dense_matmul(b, sp_a,
￼                                                            adjoint_a= not adjoin_a,
￼                                                            adjoint_b=not adjoin_b))
```

Comment probably needs rewrite, but code will be clearer, I think

joglekara replied:
That looks great. Thanks for that!

mihaimaruseac commented on tensorflow/python/ops/sparse_ops_test.py:
```diff
@@ -30,6 +30,7 @@
 from tensorflow.python.ops import array_grad  # pylint: disable=unused-import
 from tensorflow.python.ops import gradient_checker_v2 as gradient_checker
 from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import array_ops
```
This should be moved below after the line importing `array_grad` (as the imports have to be sorted)

joglekara replied:
Done, thanks!

Commit:
Moved the import statement to conform to alphabetical rules

Commit:
and consolidated if-statements thanks to review

joglekara commented:
@mihaimaruseac, I made both of the proposed changes. Feel free to take another look.

mihaimaruseac reviewed (dismissed):
Make sure to run pylint as I think some lines are now over 80 characters. But it looks good.

Thank you

Commit:
linter

joglekara commented:
Gotcha. Just ran the linter. Let's try again
